src/cc/BPFTable.h

```diff
@@ -36,42 +36,83 @@ namespace ebpf {
 template <class KeyType, class ValueType>
 class BPFTableBase {
  public:
-  size_t capacity() { return capacity_; }
+  size_t capacity() { return desc.max_entries; }
+
+  StatusTuple string_to_key(const std::string &key_str, KeyType *key) {
+    if (bpf_module->table_key_scanf(id, key_str.c_str(), key))
+      return StatusTuple(-1, "Error on key_sscanff: %s", std::strerror(errno));
+    return StatusTuple(0);
+  }
+
+  StatusTuple string_to_leaf(const std::string &value_str, ValueType *value) {
+    if (bpf_module->table_leaf_scanf(id, value_str.c_str(), value))
+      return StatusTuple(-1, "Error on leaf_sscanff: %s", std::strerror(errno));
+    return StatusTuple(0);
+  }
+
+  StatusTuple key_to_string(const KeyType *key, std::string &key_str) {
+    char buf[8 * desc.key_size];
+    if (bpf_module->table_key_printf(id, buf, sizeof(buf), key))
+      return StatusTuple(-1, "Error on key_sprintf: %s", std::strerror(errno));
+
+    key_str.assign(buf);
+    return StatusTuple(0);
+  }
+
+  StatusTuple leaf_to_string(const ValueType *value, std::string &value_str) {
+    char buf[8 * desc.leaf_size];
+    if (bpf_module->table_leaf_printf(id, buf, sizeof(buf), value))
+      return StatusTuple(-1, "Error on leaf_sprintf: %s", std::strerror(errno));
+
+    value_str.assign(buf);
+    return StatusTuple(0);
+  }
 
  protected:
-  explicit BPFTableBase(const TableDesc& desc) {
-    fd_ = desc.fd;
-    capacity_ = desc.max_entries;
+  explicit BPFTableBase(const TableDesc &desc, BPFModule *bpf_module) :
+    desc(desc), bpf_module(bpf_module) {
+    id = bpf_module->table_id(desc.name);
   }
 
   bool lookup(KeyType* key, ValueType* value) {
-    return bpf_lookup_elem(fd_, static_cast<void*>(key),
+    return bpf_lookup_elem(desc.fd, static_cast<void*>(key),
                            static_cast<void*>(value)) >= 0;
   }
 
   bool next(KeyType* key, KeyType* next_key) {
-    return bpf_get_next_key(fd_, static_cast<void*>(key),
+    return bpf_get_next_key(desc.fd, static_cast<void*>(key),
                             static_cast<void*>(next_key)) >= 0;
   }
 
   bool update(KeyType* key, ValueType* value) {
-    return bpf_update_elem(fd_, static_cast<void*>(key),
+    return bpf_update_elem(desc.fd, static_cast<void*>(key),
                            static_cast<void*>(value), 0) >= 0;
   }
 
   bool remove(KeyType* key) {
-    return bpf_delete_elem(fd_, static_cast<void*>(key)) >= 0;
+    return bpf_delete_elem(desc.fd, static_cast<void*>(key)) >= 0;
   }
 
-  int fd_;
-  size_t capacity_;
+  const TableDesc &desc;
+  BPFModule *bpf_module;
+  size_t id; // table id inside bpf_module
```

Comment:
I think in the long run we want to eliminate the Table ID thing (which is essentially just index of the table in one particular `BPFModule`, and doesn't make so much sense globally), see discussions in #1040. 

Can you move the `scanf` / `printf` logic into `TableDesc` and just call that here? `BPFModule` can probably also just proxy the call for backward compatibility.

@drzaeus77 what do you think?